_posts/2021-04-20-custom-errors.md

```diff
@@ -0,0 +1,150 @@
+
+contract TestToken {
+    mapping(address => uint) balance;
+    function transfer(address to, uint256 amount) public {
+        if (amount > balance[msg.sender])
+            // Error call using named parameters. Equivalent to
+            // InsufficientBalance(balance[msg.sender], amount)
+            revert InsufficientBalance({
+                available: balance[msg.sender],
+                required: amount
+            });
+        balance[msg.sender] -= amount;
+        balance[to] += amount;
+    }
+    // ...
+}
+```
+
+The error data would be encoded identically as the ABI encoding for function calls, i.e.,
+``abi.encodeWithSignature("InsufficientBalance(uint256,uint256)", balance[msg.sender], amount)``.
+
+We hope that frameworks will provide support for Errors. The following is an example on decoding
+error data using [ethers.js](https://docs.ethers.io/v5/):
+
+```javascript
+import { ethers } from "ethers";
+
+// As a workaround, we have a function with the same name and type as the error in the abi
+const abi = [{
+    "inputs": [{
+            "internalType": "uint256",
+            "name": "available",
+            "type": "uint256"
+        },
+        {
+            "internalType": "uint256",
+            "name": "required",
+            "type": "uint256"
+        }
+    ],
+    "name": "InsufficientBalance",
+    "outputs": [],
+    "stateMutability": "nonpayable",
+    "type": "function"
+}];
+
+const interface = new ethers.utils.Interface(abi);
+// Error data obtained after calling `TestToken.transfer(...)`.
+const error_data = "0xcf47918100000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000100000000";
+interface.decodeFunctionData(interface.functions['InsufficientBalance(uint256,uint256)'], error_data);
+// [
+//   BigNumber { _hex: '0x0100', _isBigNumber: true },
+//   BigNumber { _hex: '0x0100000000', _isBigNumber: true },
+//   available: BigNumber { _hex: '0x0100', _isBigNumber: true },
+//   required: BigNumber { _hex: '0x0100000000', _isBigNumber: true }
+// ]
+```
+
+The compiler includes all errors that a contract can emit in the contract's
+[ABI-JSON](https://docs.soliditylang.org/en/latest/abi-spec.html?#json). Note that this will not
+include errors emitted through external calls. Similarly, developers can provide
+[NatSpec](https://docs.soliditylang.org/en/latest/natspec-format.html) documentation for Errors
+which would then be part of the user and developer documentation.
+
+Note that one can never trust the error data. The error data by default bubbles up through the chain
+of external calls, which means that a contract may receive an error not defined in any of the
+contracts it calls directly. Furthermore, any contract can fake any error by returning data that
+matches an error signature, even if the error is not defined anywhere.
+
+Currently, there is no convenient way to catch errors in Solidity, but this is planned.
```

Comment:
Do we have an issue or anything that we can link to or can we make the "this is planned" more detailed or sound a bit nicer?

Reply:
I couldn't find any issue. I'll make a new one.